Documentation to existing code + fixes parameter detection on bytecode->goto translation

cesaro opened the pull request:
Fixes [test-gen/#581](https://github.com/diffblue/test-gen/issues/581).

smowton commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -1120,6 +1153,10 @@ codet java_bytecode_convert_methodt::convert_instructions(
   }
 
+  // Clean the list of temporary variables. These are local variables in the
+  // goto function used to represent temporary values of the JVM operand stack?
```
Also other things such as temporaries to hold a newly-allocated instance between allocation and a constructor being called. Any caller of `tmp_variable`, in short.

cesaro replied:
Done.

smowton commented on src/java_bytecode/java_bytecode_convert_method_class.h:
```diff
@@ -147,8 +163,18 @@ class java_bytecode_convert_methodt:public messaget
   void pop_residue(std::size_t n);
   void push(const exprt::operandst &o);
 
+  /// Determines whether the `method` is a constructor, by checking whether its
```
or a static initializer (`<clinit>`)

cesaro replied:
Done.

smowton commented on src/java_bytecode/java_local_variable_table.cpp:
```diff
@@ -186,9 +187,10 @@ static bool is_store_to_slot(
+      "expected store instruction looks like store_0, store_1...");
     storeslot=prevstatement[7];
-    assert(isdigit(storeslot[0]));
+    INVARIANT(isdigit(storeslot[0]), "digit expected");
```
Description doesn't add anything on top of the code. How about "`store_ instructions should end in a digit`"

cesaro replied:
Indeed, used your description; thanks.

smowton commented on src/java_bytecode/java_local_variable_table.cpp:
```diff
@@ -261,45 +272,72 @@ static void populate_predecessor_map(
+    // All entries of the "local_variable_table_with_holest" processed in this
+    // function concern the same Java Local Variable Table slot/register. This
+    // is because "find_initialisers()" has already sorted them.
+    INVARIANT(it->var.index == firstvar->var.index,
```
style (== spacing)

cesaro replied:
Fixed

smowton commented on src/java_bytecode/java_utils.cpp:
```diff
@@ -20,3 +22,94 @@ bool java_is_array_type(const typet &type)
                     "java::array[");
 }
+
+unsigned java_parameter_count(
```
Duplicate of https://github.com/diffblue/cbmc/blob/test-gen-support/src/java_bytecode/java_types.cpp#L162 which was used to generate the method's `code_typet` to begin with. How about making https://github.com/diffblue/cbmc/blob/test-gen-support/src/java_bytecode/java_bytecode_convert_method.cpp#L111 public and then iterating over the `code_typet`'s formal parameters?

cesaro replied:
@smowton @mgudemann, I can reimplement the function by calling indeed `java_type_from_string` and then use `count_slots` (L111) in an iteration through the arguments. Should I move the resulting reimplementation to `java_types.{h,cpp}` or I still leave it in `java_utils.{h,cpp}`?

cesaro replied:
I reimplemented the two functions and refactored the code as proposed by @smowton

mgudemann commented on src/java_bytecode/java_local_variable_table.cpp:
```diff
@@ -651,11 +696,21 @@ void java_bytecode_convert_methodt::setup_local_variables(
   method_with_amapt dominator_args(m, amap);
   dominator_analysis(dominator_args);
 
+#ifdef DEBUG
```
is this necessary when you use `debug()` anyway?

cesaro replied:
Macro `DEBUG` is by default not defined, so even if you use `--verbosity 10` you won't see these three messages unless you explicitly define it. Try to compile `cbmc` with `DEBUG` defined and you will see many other conveniently hidden messages, like these.

mgudemann commented on src/java_bytecode/java_local_variable_table.cpp:
```diff
@@ -673,23 +728,32 @@ void java_bytecode_convert_methodt::setup_local_variables(
+    if(v.is_parameter)
       continue;
 
+#ifdef DEBUG
```
as above

cesaro replied:
Ditto

mgudemann commented on src/java_bytecode/java_utils.cpp:
```diff
@@ -20,3 +22,94 @@ bool java_is_array_type(const typet &type)
+  unsigned num_slots=0;
+  bool saw_array;
+
+  INVARIANT(method_descriptor.size()>=3, "function always look like ()T");
```
"functions always have at least `()T` with `T` some type as descriptor"

mgudemann commented on src/java_bytecode/java_utils.cpp:
```diff
@@ -20,3 +22,94 @@ bool java_is_array_type(const typet &type)
+  bool saw_array;
+
+  INVARIANT(method_descriptor.size()>=3, "function always look like ()T");
+  INVARIANT(method_descriptor[0]=='(', "function always look like ()T");
```
"function descriptor always starts with `(`"

mgudemann commented on src/java_bytecode/java_utils.cpp:
```diff
@@ -20,3 +22,94 @@ bool java_is_array_type(const typet &type)
+  INVARIANT(method_descriptor.size()>=3, "function always look like ()T");
+  INVARIANT(method_descriptor[0]=='(', "function always look like ()T");
+
+  for(i=1; method_descriptor[i]!=')'; i++)
```
scope of `i` should be reduced to this loop

mgudemann commented on src/java_bytecode/java_utils.cpp:
```diff
@@ -20,3 +22,94 @@ bool java_is_array_type(const typet &type)
+    UNREACHABLE;
+    return 0;
+  }
+}
```
this seems to duplicate what already exists to get a java type from a descriptor (`java_type_from_string`), but with the constraint on function types

mgudemann commented on src/java_bytecode/java_utils.h:
```diff
@@ -9,9 +9,49 @@ Author: Daniel Kroening, kroening@kroening.com
 bool java_is_array_type(const typet &type);
 
+/// Used to indicate what to count in the `java_parameter_count` functions.
+enum java_parameter_count_whatt
```
use `enum class`

cesaro commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -297,127 +298,160 @@ void java_bytecode_convert_methodt::convert(
   method.set(ID_abstract, m.is_abstract);
   method.set(ID_is_virtual, is_virtual);
-
+  method.type()=member_type;
```
Because that line also initializes the object `method` and so it also belongs to the "initialization paragraph" above.

cesaro commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -297,127 +298,160 @@ void java_bytecode_convert_methodt::convert(
 
+  // Set up the pretty name for the method entry in the symbol table.
+  // The pretty name of a constructor includes the base name of the class
+  // instead of the internal .class name "<init>". For regular methods, it's
```
Done

cesaro commented:
- Rebased my original commits into 3 commits as suggested by @smowton 
- Ran the linter and fixed warnings on code introduced or modified by this PR, as proposed by @mgudemann 
- Rewrote the duplicated function to be based on existing (refactored) code

@smowton, @mgudemann, please review again and let me know.

smowton commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -297,127 +283,159 @@ void java_bytecode_convert_methodt::convert(
 
+  // Determine the number of local variable slots used by the JVM to maintan the
+  // formal parameters
+  slots_for_parameters = java_method_parameter_slots(code_type);
```
Style: = spacing

cesaro replied:
Ran the linter again on all my files, dind't discover anything new apart from what you spotted.

cesaro commented on src/java_bytecode/java_bytecode_convert_method_class.h:
```diff
@@ -147,8 +163,18 @@ class java_bytecode_convert_methodt:public messaget
+  /// `slots_for_parameters` is initialized upon call.
+  bool is_parameter(const local_variablet &v)
+  {
+    return v.index < slots_for_parameters;
```
Done.

cesaro commented on src/java_bytecode/java_local_variable_table.cpp:
```diff
@@ -55,7 +56,7 @@ struct procedure_local_cfg_baset<
     {
       auto findit=amap.find(table_entry.start_pc);
-      assert(findit!=amap.end() &&
+      INVARIANT(findit!=amap.end(),
```
Done.

cesaro commented on src/java_bytecode/java_local_variable_table.cpp:
```diff
@@ -186,9 +187,11 @@ static bool is_store_to_slot(
   {
     // Store shorthands, like "store_0", "store_1"
-    assert(prevstatement[6]=='_' && prevstatement.size()==8);
+    INVARIANT(prevstatement[6]=='_' && prevstatement.size()==8,
```
Did all of the invariants in `src/java_bytecode`, thank you!

peterschrammel commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -106,25 +107,10 @@ static bool operator==(const irep_idt &what, const patternt &pattern)
   const auto to_strip_str=id2string(to_strip);
-  assert(has_prefix(to_strip_str, "java::"));
+  PRECONDITION(has_prefix(to_strip_str, "java::"));
   return to_strip_str.substr(6, std::string::npos);
```
I'd define a constant string "java::" and take its length instead of using a magic number.

cesaro replied:
@peterschrammel, I dind't change that code, only the assert -> precondition, as I was getting warnings by the linter. I can do the change that you suggest, if you want.

peterschrammel commented on src/java_bytecode/java_local_variable_table.cpp:
```diff
@@ -261,45 +276,76 @@ static void populate_predecessor_map(
-    if(it->var.start_pc==0)
+    // All entries of the "local_variable_table_with_holest" processed in this
+    // function concern the same Java Local Variable Table slot/register. This
+    // is because "find_initialisers()" has already sorted them.
```
initializers

cesaro replied:
This was @smowton's code, will change this and the PRECONDITION thing above

peterschrammel commented on src/java_bytecode/java_local_variable_table.cpp:
```diff
@@ -651,11 +704,21 @@ void java_bytecode_convert_methodt::setup_local_variables(
+    vars_with_holes.push_back({v, is_parameter(v), {}});
 
   // Merge variable records:
   find_initialisers(vars_with_holes, amap, dominator_analysis);
```
rename to `find_initializers`

cesaro replied:
Done

peterschrammel commented on src/java_bytecode/java_local_variable_table.cpp:
```diff
@@ -316,10 +362,17 @@ static void populate_predecessor_map(
                it->var.index))
           {
-            throw "local variable table: didn't find initialising store";
+            msg.error() << "Local variable table: didn't find initialising "
```
initializing

cesaro replied:
Done.

peterschrammel commented on src/java_bytecode/java_local_variable_table.cpp:
```diff
@@ -261,45 +276,76 @@ static void populate_predecessor_map(
       continue;
 
+    msg.debug() << "ppm: processing var idx " << it->var.index <<
+                   " name '" << it->var.name << "' start-pc "
```
align <<

cesaro replied:
Done

cesaro commented on src/java_bytecode/java_utils.cpp:
```diff
@@ -20,3 +22,33 @@ bool java_is_array_type(const typet &type)
+
+unsigned java_method_parameter_slots(const code_typet &t)
+{
+  unsigned slots = 0;
```
Thank you, done.

peterschrammel commented on src/java_bytecode/java_local_variable_table.cpp:
```diff
@@ -261,45 +276,76 @@ static void populate_predecessor_map(
-          assert(inst_before_this!=amap.begin());
+          INVARIANT(
+            inst_before_this!=amap.begin(),
+            "we shall not be on the first bytecode of the method");
```
Suggestion (should probably go into coding guideline): Can we consistently start these messages with an upper case letter?

cesaro replied:
@peterschrammel, currently the coding style says that a "literal string" thrown as exception needs to start with lower cases. Should both rules agree on one choice?

peterschrammel replied:
I am fine with lower case as long as it is consistent.
